### PR TITLE
Staging hosts and resolv.conf files during build

### LINKED ISSUE #4579: Updating CentOS 8 container fails
### Version of Singularity:

What version of Singularity are you using? Run:

```
$ singularity version
3.4.1-1.1.el7
```

### Expected behavior

Using a recipe `centos8.def`:
```
BootStrap: docker
From: centos:centos8

%post
yum -y update
```
I expect to be able to build a container being up-to-date with the most recent security updates doing:
```
singularity build centos8.sif centos8.def
```

### Actual behavior

Errors out during `yum -y update`:
```
...
  Upgrading        : setup-2.12.2-2.el8.noarch                                                                                                                                                                                         21/38 
Error unpacking rpm package setup-2.12.2-2.el8.noarch
error: unpacking of archive failed on file /etc/hosts: cpio: utime

  Upgrading        : bind-export-libs-32:9.11.4-17.P2.el8_0.1.x86_64                                                                                                                                                                   22/38 
  Running scriptlet: bind-export-libs-32:9.11.4-17.P2.el8_0.1.x86_64                                                                                                                                                                   22/38 
error: setup-2.12.2-2.el8.noarch: install failed

  Cleanup          : systemd-udev-239-13.el8.x86_64                                                                                                                                                                                    23/38 
...
Error: Transaction failed
FATAL:   failed to execute %post proc: exit status 1
FATAL:   While performing build: while running engine: exit status 255
```

### What OS/distro are you running

```
$ lsb_release -a
LSB Version:    :core-4.1-amd64:core-4.1-noarch
Distributor ID: CentOS
Description:    CentOS Linux release 7.7.1908 (Core) 
Release:        7.7.1908
Codename:       Core
```

### How did you install Singularity

EPEL 7

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

Copy `/etc/hosts` and `/etc/resolv.conf` to a `tmpfs` directory and bind them into container instead of binding host files during image build.

### This fixes or addresses the following GitHub issues:

 - Fixes #4579 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Review by dtrudg — CHANGES_REQUESTED** [review]
This does not yet fix the issue when I use my vagrant reproducer:

```
3.4.2+459-gb963cfd
```


```
$ sudo singularity build centos8.sif centos8.def
...
Error unpacking rpm package setup-2.12.2-2.el8.noarch
error: unpacking of archive failed on file /etc/hosts: cpio: utime
...
Failed:
  setup-2.12.2-2.el8.noarch                                                             

Error: Transaction failed
FATAL:   failed to execute %post proc: exit status 1
FATAL:   While performing build: while running engine: while running /usr/local/libexec/singularity/bin/starter: exit status 255
```

I think the staged (copy) of `hosts` and `resolv.conf` are still being mounted in `ro`?

File: internal/pkg/runtime/engine/imgbuild/create_linux.go [hunk 1]
```diff
@@ -135,36 +148,48 @@ func (e *EngineOperations) CreateContainer(ctx context.Context, pid int, rpcConn
...
-		return fmt.Errorf("mount /etc/hosts failed: %s", err)
+	// copy /etc/hosts to tmpfs and bind copy into container
+	sessionHosts, err := stageFile("/etc/hosts", sessionPath)
+	if err != nil {
+		return err
+	}
+
+	dest = filepath.Join(sessionRootFs, "etc", "hosts")
+	sylog.Debugf("Mounting %s at %s\n", sessionHosts, dest)
+	if err := rpcOps.Mount(sessionHosts, dest, "", flags, ""); err != nil {
```
Comment (dtrudg) [inline internal/pkg/runtime/engine/imgbuild/create_linux.go]: flags here has `MS.RD_ONLY` given that rpm is still trying to modify our staged copy, can we just mount it `rw`? Any changes made are only relevant for the session?

File: internal/pkg/runtime/engine/imgbuild/create_linux.go [hunk 2]
```diff
@@ -135,36 +148,48 @@ func (e *EngineOperations) CreateContainer(ctx context.Context, pid int, rpcConn
...
-		return fmt.Errorf("mount /etc/resolv.conf failed: %s", err)
+	// copy /etc/resolv.conf to tmpfs and bind copy into container
+	sessionResolv, err := stageFile("/etc/resolv.conf", sessionPath)
+	if err != nil {
+		return err
+	}
+
+	dest = filepath.Join(sessionRootFs, "etc", "resolv.conf")
+	sylog.Debugf("Mounting %s at %s\n", sessionResolv, dest)
+	if err := rpcOps.Mount(sessionResolv, dest, "", flags, ""); err != nil {
```
Comment (dtrudg) [inline internal/pkg/runtime/engine/imgbuild/create_linux.go]: flags here has `MS.RD_ONLY` given that rpm is still trying to modify our staged copy, can we just mount it `rw`? Any changes made are only relevant for the session?

File: internal/pkg/runtime/engine/imgbuild/create_linux.go (hunk 1 above)
Comment (cclerget) [reply internal/pkg/runtime/engine/imgbuild/create_linux.go]: Oops, fixed

**Comment by dtrudg** [comment]
Tested and this is a fix....

```
[vagrant@localhost centos8]$ sudo singularity build centos8.sif centos8.def 
INFO:    Starting build...
Getting image source signatures
...
Upgraded:
  bash-4.4.19-8.el8_0.x86_64              bind-export-libs-32:9.11.4-17.P2.el8_0.1.x86_64  
  iptables-libs-1.8.2-9.el8_0.1.x86_64    kmod-25-11.el8_0.2.x86_64                        
  kmod-libs-25-11.el8_0.2.x86_64          libdnf-0.22.5-5.el8_0.x86_64                     
  p11-kit-0.23.14-5.el8_0.x86_64          p11-kit-trust-0.23.14-5.el8_0.x86_64             
  platform-python-3.6.8-4.el8_0.x86_64    python3-hawkey-0.22.5-5.el8_0.x86_64             
  python3-libdnf-0.22.5-5.el8_0.x86_64    python3-libs-3.6.8-4.el8_0.x86_64                
  python3-rpm-4.14.2-11.el8_0.x86_64      rpm-4.14.2-11.el8_0.x86_64                       
  rpm-build-libs-4.14.2-11.el8_0.x86_64   rpm-libs-4.14.2-11.el8_0.x86_64                  
  setup-2.12.2-2.el8.noarch               systemd-239-13.el8_0.5.x86_64                    
  systemd-libs-239-13.el8_0.5.x86_64      systemd-pam-239-13.el8_0.5.x86_64                
  systemd-udev-239-13.el8_0.5.x86_64     

Installed:
  libxkbcommon-0.8.2-1.el8.x86_64            diffutils-3.6-5.el8.x86_64                   
  kbd-2.0.4-8.el8.x86_64                     xkeyboard-config-2.24-3.el8.noarch           
  kbd-legacy-2.0.4-8.el8.noarch              kbd-misc-2.0.4-8.el8.noarch                  

Complete!
INFO:    Creating SIF file...
INFO:    Build complete: centos8.sif
```

**Review by dtrudg — APPROVED** [review]

**Pull request merged**